src/coprocessor/dag/batch/executors/util/heap.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cell::RefCell;
+use std::cmp::{self, Ordering};
+use std::collections::BinaryHeap;
+use std::sync::Arc;
+use std::usize;
+
+use crate::coprocessor::codec::batch::LazyBatchColumnVec;
+use crate::coprocessor::dag::expr::Result;
+
+const HEAP_MAX_CAPACITY: usize = 1024;
+
+pub struct SortRow {
+    table: Arc<RefCell<LazyBatchColumnVec>>,
+    idx: usize,
+    //(col_id,desc)
+    order_cols: Arc<Vec<(usize, bool)>>,
+}
+
+impl SortRow {
+    pub fn new(
+        table: Arc<RefCell<LazyBatchColumnVec>>,
+        idx: usize,
+        order_cols: Arc<Vec<(usize, bool)>>,
+    ) -> SortRow {
+        SortRow {
+            table,
+            idx,
+            order_cols,
+        }
+    }
+}
+
+/// TopNHeap wrapping a mutable reference to the greatest row on
+/// a heap.
+pub struct TopNHeap {
+    pub rows: BinaryHeap<SortRow>,
+    pub table: Arc<RefCell<LazyBatchColumnVec>>,
+    limit: usize,
+    //(col_id,desc)
+    order_cols: Arc<Vec<(usize, bool)>>,
+}
+
+impl TopNHeap {
+    /// Create an new `TopNHeap`
+    pub fn new(
+        limit: usize,
+        data: LazyBatchColumnVec,
+        order_cols: Arc<Vec<(usize, bool)>>,
+    ) -> Result<TopNHeap> {
+        if limit == usize::MAX || limit == 0 {
+            return Err(box_err!("invalid limit:{}", limit));
+        }
+        let cap = cmp::min(limit, HEAP_MAX_CAPACITY);
+        let binary_heap = BinaryHeap::with_capacity(cap);
+
+        let current_rows = data.rows_len();
+
+        let mut heap = TopNHeap {
+            rows: binary_heap,
+            limit,
+            table: Arc::new(RefCell::new(data)),
+            order_cols,
+        };
+
+        heap.push_logical_rows(current_rows);
+        Ok(heap)
+    }
+
+    /// Pushes a batch into the binary heap.
+    pub fn push_batch(&mut self, mut data: LazyBatchColumnVec) {
+        if data.rows_len() == 0 {
+            return;
+        }
+        let data_rows = data.rows_len();
+        self.table.borrow_mut().append(&mut data);
+        self.push_logical_rows(data_rows);
+    }
+
+    /// Consumes the `TopNHeap` and returns a `LazyBatchColumnVec` in sorted order.
+    pub fn into_sorted_result(self) -> LazyBatchColumnVec {
+        let mut sorted_rows: Vec<usize> =
+            self.rows.into_sorted_vec().iter().map(|r| r.idx).collect();
+        let mut table = self.table.as_ref().borrow().to_owned();
+        let rows_len = sorted_rows.len();
+        for row_id in 0..rows_len {
+            while row_id != sorted_rows[row_id] {
+                let dest_row = sorted_rows[row_id];
+                sorted_rows.swap(row_id, dest_row);
```

Comment:
Don't swap. Instead, deconstruct the original one and then move data to the new container in the correct order one by one sequentially.